hal/armv7a/imx6ull/imx6ull.c

```diff
@@ -0,0 +1,360 @@
+	mmdc_mpwrdlhwst0 = 0x21c, mmdc_mpwlhwerr = 0x21e, mmdc_mpdghwst0, mmdc_mpdghwst1, mmdc_mpdghwst2, mmdc_mpdghwst3,
+	mmdc_mppdcmpr1, mmdc_mppdcmpr2, mmdc_mpswdar0, mmdc_mpswdrdr0, mmdc_mpswdrdr1, mmdc_mpswdrdr2, mmdc_mpswdrdr3,
+	mmdc_mpswdrdr4, mmdc_mpswdrdr5, mmdc_mpswdrdr6, mmdc_mpswdrdr7, mmdc_mpmur0, mmdc_mpwrcadl, mmdc_mpdccr
+};
+
+
+/* Reserved slots */
+static const char ccm_reserved[] = { clk_asrc + 1, clk_ipsync_ip2apb_tzasc1_ipg + 1, clk_pxp + 1,
+	clk_mmdc_core_aclk_fast_core_p0 + 1, clk_iomux_snvs_gpr + 1, clk_usdhc2 + 1 };
+
+
+struct {
+	volatile u32 *ccm;
+	volatile u32 *ccm_analog;
+
+	volatile u32 *iomux;
+	volatile u32 *iomux_gpr;
+	volatile u32 *iomux_snvs;
+
+	volatile u32 *mmdc;
+} imx6ull_common;
+
+
+static int imx6ull_isValidDev(int dev)
+{
+	int i;
+
+	if (dev < clk_aips_tz1 || dev > clk_pwm7)
+		return 0;
+
+	for (i = 0; i < sizeof(ccm_reserved) / sizeof(ccm_reserved[0]); ++i) {
+		if (dev == ccm_reserved[i])
+			return 0;
+	}
+
+	return 1;
+}
+
+
+int imx6ull_setDevClock(int dev, unsigned int state)
+{
+	int ccgr, flag, mask;
+	u32 r;
+
+	if (!imx6ull_isValidDev(dev))
+		return -1;
+
+	ccgr = dev / 16;
+	flag = (state & 3) << (2 * (dev % 16));
+	mask = 3 << (2 * (dev % 16));
+
+	r = *(imx6ull_common.ccm + ccm_ccgr0 + ccgr);
+	*(imx6ull_common.ccm + ccm_ccgr0 + ccgr) = (r & ~mask) | flag;
+
+	return 0;
+}
+
+
+int imx6ull_setIOmux(int mux, char sion, char mode)
+{
+	volatile u32 *base = imx6ull_common.iomux;
+
+	if (mux >= mux_boot_mode0 && mux <= mux_tamper9) {
+#ifdef CPU_IMX6UL
+		mux = (mux - mux_boot_mode0) + 5;
+#else
+		mux = (mux - mux_boot_mode0);
+		base = imx6ull_common.iomux_snvs;
+#endif
+	}
+	else if (mux < mux_jtag_mod || mux > mux_csi_d7)
+		return -1;
+
+	*(base + mux) = (!!sion << 4) | (mode & 0xf);
+
+	return 0;
+}
+
+
+int imx6ull_setIOpad(int pad, char hys, char pus, char pue, char pke, char ode, char speed, char dse, char sre)
+{
+	u32 t;
+	volatile u32 *base = imx6ull_common.iomux;
+
+	if (pad >= pad_test_mode && pad <= pad_tamper9) {
+#ifdef CPU_IMX6UL
+		pad = (pad - pad_test_mode) + 163;
+#else
+		pad = pad - pad_test_mode + 12;
+		base = imx6ull_common.iomux_gpr;
+#endif
+	}
+	else if (pad < pad_jtag_mod || pad > pad_csi_d7)
+		return -1;
+
+	t = (!!hys << 16) | ((pus & 0x3) << 14) | (!!pue << 13) | (!!pke << 12);
+	t |= (!!ode << 11) | ((speed & 0x3) << 6) | ((dse & 0x7) << 3) | !!sre;
+	*(base + pad) = t;
+
+	return 0;
+}
+
+
+int imx6ull_setIOisel(int isel, char daisy)
+{
+	if (isel < isel_anatop || isel > isel_usdhc2_wp)
+		return -1;
+
+	*(imx6ull_common.iomux + isel) = daisy & 0x7;
+
+	return 0;
+}
+
+
+static void imx6ull_ddrInit(void)
+{
+	*(imx6ull_common.mmdc + mmdc_mdctl) |= 0x80000000;
+
+	/* Enable DDR clock */
+	*(imx6ull_common.ccm + ccm_ccgr3) |= 0x0f300000;
+
+	/* Enable AIPSTZ 1,2,3 clocks */
+	*(imx6ull_common.ccm + ccm_ccgr0) |= 0x00000005;
+	*(imx6ull_common.ccm + ccm_ccgr6) |= 0x00040000;
+
+	/* Enable EPITs clocks */
+	*(imx6ull_common.ccm + ccm_ccgr1) |= 0x00005000;
```

Comment:
Should this be in DDR init?

Reply:
In the kernel's code:
https://github.com/phoenix-rtos/phoenix-rtos-kernel/blob/def017c456d991e0546de388827e6437f4ddc9d7/hal/armv7a/imx6ull/_init.S#L186-L201
it has been done in ddr initialization part which run before startup code.

Good note, I think that `AIPSTC` clocks should be placed before `AIPS bridge initialization` in `_init.S`. However, `EPITs` clocks might be set after setting `PLLs.` What do you think?

Reply:
Sounds good, they don't need to be set next to eachother